### DISCUSSION TITLE
Compiling static ARM64 binary for Darwin (macOS)?

### DISCUSSION THREAD

**Discussion by luckman212** [esnet/iperf#1596] [General]
Is it possible to compile a static binary of iperf3 3.15 for macOS ARM64? I'd like to be able to have a portable binary that I can drop onto a machine with just a single file.

When I downloaded the source and ran `./configure; make` I ended up with a shell wrapper script around a directory full of library files.

I read the [FAQ](https://software.es.net/iperf/faq.html) but trying to execute `./configure --enable-static-bin` or `./configure "LDFLAGS=--static" --disable-shared` on my M1 system resulted in

```
~/Downloads/iperf-3.15 $ ./configure --enable-static-bin
checking for a BSD-compatible install... /opt/homebrew/bin/ginstall -c
checking whether build environment is sane... yes
checking for a race-free mkdir -p... /opt/homebrew/bin/gmkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking whether make supports nested variables... (cached) yes
checking build system type... aarch64-apple-darwin23.1.0
checking host system type... aarch64-apple-darwin23.1.0
checking how to print strings... printf
checking whether make supports the include directive... yes (GNU style)
checking for gcc... gcc
checking whether the C compiler works... no
configure: error: in `/Users/luke/Downloads/iperf-3.15':
configure: error: C compiler cannot create executables
See `config.log' for more details
```

config.log excerpt:
```
configure:4048: $? = 1
configure:4068: checking whether the C compiler works
configure:4090: gcc    --static conftest.c  >&5
ld: unknown options: --static 
clang: error: linker command failed with exit code 1 (use -v to see invocation)
configure:4094: $? = 1
configure:4134: result: no
```

I have the latest Xcode, CLT and gcc installed via Homebrew

```
$ gcc -v
Apple clang version 15.0.0 (clang-1500.0.40.1)
Target: arm64-apple-darwin23.1.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

I also tried putting gcc-13 in my `$PATH` before `/usr/bin/gcc` to ensure that the build script used that, by adding a symlink:
```
ln -s /opt/homebrew/bin/gcc-13 /usr/local/bin/gcc
```
(No change)

stuck! 🤷‍♂️

**Comment by bmah888** [comment]
I don't have a solution for you, but I confirm the same behavior using the normal Xcode C compiler on an Intel Mac (MacBookPro16,1 and macOS 14.1.1). I'm a little surprised this doesn't work.

**Comment by DarkDust** [comment]
While it might be possible to create static binaries on macOS in theory, in practice it's [deliberately NOT supported by Apple](https://developer.apple.com/library/archive/qa/qa1118/_index.html).

**Comment by luckman212** [comment]
Maybe I've got my terminology wrong, but my question is really: is it possible to create a binary for macOS using the latest 3.16 code such that a single monolithic file is all that's needed to run the program?

Similar to the ancient 3.1.3 version that's currently listed at iperf.fr

[iPerf 3.1.3](https://iperf.fr/download/apple/iperf-3.1.3-macos-x86_64.zip)
